docs/man/man1/gdal_proximity.1
.TH "gdal_proximity" 1 "Mon Nov 20 2017" "GDAL" \" -*- nroff -*-
.ad l
.nh
.SH NAME
gdal_proximity \- Produces a raster proximity map\&.
.SH "SYNOPSIS"
.PP
.PP
.nf
gdal_proximity.py srcfile dstfile [-srcband n] [-dstband n]
                  [-of format] [-co name=value]*
                  [-ot Byte/Int16/Int32/Float32/etc]
                  [-values n,n,n] [-distunits PIXEL/GEO]
                  [-maxdist n] [-nodata n] [-use_input_nodata YES/NO]
                  [-fixed-buf-val n]
.fi
.PP
.SH "DESCRIPTION"
.PP
The gdal_proximity\&.py script generates a raster proximity map indicating the distance from the center of each pixel to the center of the nearest pixel identified as a target pixel\&. Target pixels are those in the source raster for which the raster pixel value is in the set of target pixel values\&.
.PP
.IP "\fB\fIsrcfile\fP\fP" 1c
The source raster file used to identify target pixels\&.
.PP
.IP "\fB\fIdstfile\fP\fP" 1c
The destination raster file to which the proximity map will be written\&. It may be a pre-existing file of the same size as srcfile\&. If it does not exist it will be created\&.
.PP
.IP "\fB\fB-srcband\fP \fIn\fP\fP" 1c
Identifies the band in the source file to use (default is 1)\&.
.PP
.IP "\fB\fB-dstband\fP \fIn\fP\fP" 1c
Identifies the band in the destination file to use (default is 1)\&.
.PP
.IP "\fB\fB-of\fP \fIformat\fP:\fP" 1c
Select the output format\&. The default is GeoTIFF (GTiff)\&. Use the short format name\&. 
.PP
.IP "\fB\fB-co\fP \fI'NAME=VALUE'\fP:\fP" 1c
passes a creation option to the output format driver\&. Multiple \fB-co\fP options may be listed\&. See format specific documentation for legal creation options for each format\&. 
.PP
.IP "\fB\fB-ot\fP \fIdatatype\fP:\fP" 1c
Force the output image bands to have a specific type\&. Use type names (i\&.e\&. Byte, Int16,\&.\&.\&.) 
.PP
.IP "\fB\fB-values\fP \fIn,n,n\fP:\fP" 1c
A list of target pixel values in the source image to be considered target pixels\&. If not specified, all non-zero pixels will be considered target pixels\&. 
.PP
.IP "\fB\fB-distunits\fP \fIPIXEL/GEO\fP:\fP" 1c
Indicate whether distances generated should be in pixel or georeferenced coordinates (default PIXEL)\&. 
.PP
.IP "\fB\fB-maxdist\fP \fIn\fP:\fP" 1c
The maximum distance to be generated\&. The nodata value will be used for pixels beyond this distance\&. If a nodata value is not provided, the output band will be queried for its nodata value\&. If the output band does not have a nodata value, then the value 65535 will be used\&. Distance is interpreted in pixels unless -distunits GEO is specified\&. 
.PP
.IP "\fB\fB-nodata\fP \fIn\fP:\fP" 1c
Specify a nodata value to use for the destination proximity raster\&. 
.PP
.IP "\fB\fB-use_input_nodata\fP \fIYES/NO\fP:\fP" 1c
(GDAL >= 2\&.0) Indicate whether nodata pixels in the input raster should be nodata in the output raster (default NO)\&. 
.PP
.IP "\fB\fB-fixed-buf-val\fP \fIn\fP:\fP" 1c
Specify a value to be applied to all pixels that are within the -maxdist of target pixels (including the target pixels) instead of a distance value\&. 
.PP
.PP
.SH "AUTHORS"
.PP
Frank Warmerdam warmerdam@pobox.com 


bindings/python/scripts/gdal_proximity.py
#!/usr/bin/env python
#******************************************************************************
#  $Id: gdal_proximity.py 33790 2016-03-26 12:42:12Z goatbar $
#
#  Name:     gdalproximity
#  Project:  GDAL Python Interface
#  Purpose:  Application for computing raster proximity maps.
#  Author:   Frank Warmerdam, warmerdam@pobox.com
#
#******************************************************************************
#  Copyright (c) 2008, Frank Warmerdam
#  Copyright (c) 2009-2011, Even Rouault <even dot rouault at mines-paris dot org>
#
#  Permission is hereby granted, free of charge, to any person obtaining a
#  copy of this software and associated documentation files (the "Software"),
#  to deal in the Software without restriction, including without limitation
#  the rights to use, copy, modify, merge, publish, distribute, sublicense,
#  and/or sell copies of the Software, and to permit persons to whom the
#  Software is furnished to do so, subject to the following conditions:
#
#  The above copyright notice and this permission notice shall be included
#  in all copies or substantial portions of the Software.
#
#  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
#  OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
#  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
#  THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
#  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
#  FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
#  DEALINGS IN THE SOFTWARE.
#******************************************************************************

import sys

from osgeo import gdal

def Usage():
    print("""
gdal_proximity.py srcfile dstfile [-srcband n] [-dstband n]
                  [-of format] [-co name=value]*
                  [-ot Byte/Int16/Int32/Float32/etc]
                  [-values n,n,n] [-distunits PIXEL/GEO]
                  [-maxdist n] [-nodata n] [-use_input_nodata YES/NO]
                  [-fixed-buf-val n] [-q] """)
    sys.exit(1)

# =============================================================================
#     Mainline
# =============================================================================

format = 'GTiff'
creation_options = []
options = []
src_filename = None
src_band_n = 1
dst_filename = None
dst_band_n = 1
creation_type = 'Float32'
quiet_flag = 0

gdal.AllRegister()
argv = gdal.GeneralCmdLineProcessor( sys.argv )
if argv is None:
    sys.exit( 0 )

# Parse command line arguments.
i = 1
while i < len(argv):
    arg = argv[i]

    if arg == '-of':
        i = i + 1
        format = argv[i]

    elif arg == '-co':
        i = i + 1
        creation_options.append(argv[i])

    elif arg == '-ot':
        i = i + 1
        creation_type = argv[i]

    elif arg == '-maxdist':
        i = i + 1
        options.append( 'MAXDIST=' + argv[i] )

    elif arg == '-values':
        i = i + 1
        options.append( 'VALUES=' + argv[i] )

    elif arg == '-distunits':
        i = i + 1
        options.append( 'DISTUNITS=' + argv[i] )

    elif arg == '-nodata':
        i = i + 1
        options.append( 'NODATA=' + argv[i] )

    elif arg == '-use_input_nodata':
        i = i + 1
        options.append( 'USE_INPUT_NODATA=' + argv[i] )

    elif arg == '-fixed-buf-val':
        i = i + 1
        options.append( 'FIXED_BUF_VAL=' + argv[i] )

    elif arg == '-srcband':
        i = i + 1
        src_band_n = int(argv[i])

    elif arg == '-dstband':
        i = i + 1
        dst_band_n = int(argv[i])

    elif arg == '-q' or arg == '-quiet':
        quiet_flag = 1

    elif src_filename is None:
        src_filename = argv[i]

    elif dst_filename is None:
        dst_filename = argv[i]

    else:
        Usage()

    i = i + 1

if src_filename is None or dst_filename is None:
    Usage()

# =============================================================================
#    Open source file
# =============================================================================

src_ds = gdal.Open( src_filename )

if src_ds is None:
    print('Unable to open %s' % src_filename)
    sys.exit(1)

srcband = src_ds.GetRasterBand(src_band_n)

# =============================================================================
#       Try opening the destination file as an existing file.
# =============================================================================

try:
    driver = gdal.IdentifyDriver( dst_filename )
    if driver is not None:
        dst_ds = gdal.Open( dst_filename, gdal.GA_Update )
        dstband = dst_ds.GetRasterBand(dst_band_n)
    else:
        dst_ds = None
except:
    dst_ds = None

# =============================================================================
#     Create output file.
# =============================================================================
if dst_ds is None:
    drv = gdal.GetDriverByName(format)
    dst_ds = drv.Create( dst_filename,
                         src_ds.RasterXSize, src_ds.RasterYSize, 1,
                         gdal.GetDataTypeByName(creation_type), creation_options )

    dst_ds.SetGeoTransform( src_ds.GetGeoTransform() )
    dst_ds.SetProjection( src_ds.GetProjectionRef() )

    dstband = dst_ds.GetRasterBand(1)

# =============================================================================
#    Invoke algorithm.
# =============================================================================

if quiet_flag:
    prog_func = None
else:
    prog_func = gdal.TermProgress

gdal.ComputeProximity( srcband, dstband, options,
                       callback = prog_func )

srcband = None
dstband = None
src_ds = None
dst_ds = None
